lib/_http_common.js

```diff
@@ -237,57 +237,15 @@ function checkIsHttpToken(val) {
   return tokenRegExp.test(val);
 }
 
+const headerCharRegex = /^[\t\x20-\x7e\x80-\xff]*$/;
 /**
  * True if val contains an invalid field-vchar
  *  field-value    = *( field-content / obs-fold )
  *  field-content  = field-vchar [ 1*( SP / HTAB ) field-vchar ]
  *  field-vchar    = VCHAR / obs-text
- *
- * checkInvalidHeaderChar() is currently designed to be inlinable by v8,
- * so take care when making changes to the implementation so that the source
- * code size does not exceed v8's default max_inlined_source_size setting.
  **/
-var validHdrChars = [
-  0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, // 0 - 15
-  0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, // 16 - 31
-  1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, // 32 - 47
-  1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, // 48 - 63
-  1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, // 64 - 79
-  1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, // 80 - 95
-  1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, // 96 - 111
-  1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, // 112 - 127
-  1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, // 128 ...
-  1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
-  1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
-  1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
-  1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
-  1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
-  1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
-  1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1  // ... 255
-];
 function checkInvalidHeaderChar(val) {
-  val += '';
-  if (val.length < 1)
-    return false;
-  if (!validHdrChars[val.charCodeAt(0)])
-    return true;
-  if (val.length < 2)
-    return false;
-  if (!validHdrChars[val.charCodeAt(1)])
-    return true;
-  if (val.length < 3)
-    return false;
-  if (!validHdrChars[val.charCodeAt(2)])
-    return true;
-  if (val.length < 4)
-    return false;
-  if (!validHdrChars[val.charCodeAt(3)])
-    return true;
-  for (var i = 4; i < val.length; ++i) {
-    if (!validHdrChars[val.charCodeAt(i)])
-      return true;
-  }
-  return false;
+  return !headerCharRegex.test(val);
```

Comment:
1. Rather than doing `!/^[...]*$/.test(val)`, it might be simpler to do `/[^...]/.test(val)`.
2. It seems there are some regressions in the benchmark. If they still exist after swapping the conditions as I did above, we might want to add a fast case for things like the empty string.

Reply:
I am relatively certain that there will still be a certain amount of regression even with that change. If not: wonderful. But in case the regressions continue to exist: I would like to keep the implementation and add the `RegExp` for `val.length > 5`.

Reply:
Sorry for the delayed response, and thanks for the feedback!

1. Agreed, `/[^...]/.test(val)` is prettier. I'll update it.
2. I think regressing some particular strings is okay if the expected overall throughput on realistic data is improved. I don't have real-world data, but I do have data from an AcmeAir run (which uses popular modules in relatively straightforward ways, so might be somewhat representative of real data), and it tends to call this method with longer strings where the regex is faster. The same sort of suggestions came up in the [previous change to simplify checkIsHttpToken](https://github.com/nodejs/node/pull/17399), and people ended up deciding that the readability win was worth regressing a few benchmark cases. Furthermore, V8 folks like @bmeurer and @psmarshall are committed to ensuring that idiomatic code is performant, and there is already an [open bug](https://bugs.chromium.org/p/v8/issues/detail?id=7200) for improving regex performance on short inputs.

Reply:
Comparison between iterations (old is `!/^[...]*$/.test(val)`, new is `/[^...]/.test(val)`):

```
                                                                                                improvement confidence      p.value
 http\\check_invalid_header_char.js n=1000000 input="\177"                                        -10.43 %        *** 1.145512e-11
 http\\check_invalid_header_char.js n=1000000 input=""                                             20.55 %        *** 5.135039e-20
 http\\check_invalid_header_char.js n=1000000 input="\\t\\t\\t\\t\\t\\t\\t\\t\\t\\tFoo bar baz"    -1.96 %            3.430864e-01
 http\\check_invalid_header_char.js n=1000000 input="1"                                            18.08 %        *** 1.373559e-22
 http\\check_invalid_header_char.js n=1000000 input="20091"                                        13.54 %        *** 4.766386e-09
 http\\check_invalid_header_char.js n=1000000 input="ä,-æ-╪å`¢"                                    -9.09 %        *** 2.598671e-09
 http\\check_invalid_header_char.js n=1000000 input="close"                                        14.75 %        *** 5.152685e-10
 http\\check_invalid_header_char.js n=1000000 input="en-US"                                        15.73 %        *** 4.681282e-09
 http\\check_invalid_header_char.js n=1000000 input="foo\\nbar"                                    -1.83 %            1.630593e-01
 http\\check_invalid_header_char.js n=1000000 input="group_acmeair"                               -13.64 %        *** 8.914585e-15
 http\\check_invalid_header_char.js n=1000000 input="gzip"                                         14.60 %        *** 7.965079e-26
 http\\check_invalid_header_char.js n=1000000 input="Here is a value that is real..."(truncated)   55.78 %        *** 9.914895e-30
 http\\check_invalid_header_char.js n=1000000 input="keep-alive"                                   10.51 %        *** 5.515997e-09
 http\\check_invalid_header_char.js n=1000000 input="private"                                       9.98 %        *** 4.858798e-09
 http\\check_invalid_header_char.js n=1000000 input="SAMEORIGIN"                                    7.52 %        *** 1.855577e-13
 http\\check_invalid_header_char.js n=1000000 input="Sat, 07 May 2016 16:54:48 GMT"                -7.75 %        *** 1.633302e-04
 http\\check_invalid_header_char.js n=1000000 input="text/html; charset=utf-8"                     -8.10 %        *** 2.957433e-07
 http\\check_invalid_header_char.js n=1000000 input="text/plain"                                    7.21 %        *** 2.673265e-11
```

All of these changes are pretty small compared to the overall numbers. I'll update the table in the description as soon as I push the updated code.